### PR TITLE
Experimental new program interface

### PR DESCRIPTION
**Pull request by IceKhan13**
### Summary

POC of new interface for program. It adds ability to define and run program inside one file, which is useful for notebook execution.

### Details and comments

```python
from quantum_serverless import get_arguments, save_result, distribute_task, distribute_program, get

from qiskit import QuantumCircuit
from qiskit.primitives import Sampler
from qiskit.circuit.random import random_circuit

@distribute_task(target={"cpu": 2})
def distributed_sample(circuit: QuantumCircuit):
    """Distributed task that returns quasi distribution for given circuit."""
    return Sampler().run(circuit).result().quasi_dists[0]


@distribute_program(provider=Provider(...))
def program_with_distributed_tasks(circuits):
    sample_task_references = [
        distributed_sample(circuit)
        for circuit in circuits
    ]
    results = get(sample_task_references)
    print(results)


circuits = []
for _ in range(3):
    circuit = random_circuit(2, 2)
    circuit.measure_all()
    circuits.append(circuit)

job = program_with_distributed_tasks(circuits=circuits)
job.result()
print(job.logs())
```

Notebook -> https://github.com/Qiskit-Extensions/quantum-serverless/blob/feature/program-decorator/docs/running/notebooks/06_running_programs_using_decorators.ipynb

### CODE REVIEW

File: client/quantum_serverless/core/decorators.py [hunk 1]
```diff
@@ -317,3 +321,116 @@ def wrapper(*args, **kwargs):
...
+
+def distribute_program(
+    provider: Optional[Any] = None,
+    dependencies: Optional[List[str]] = None,
+    working_dir: Optional[str] = None,
+):
+    """[Experimental] Program decorator to turn function into remotely executable program.
+
+    Example:
+        >>> @program(provider=Provider(...), dependencies=[...])
```
Comment (akihikokuroda) [inline client/quantum_serverless/core/decorators.py:349]: @distributed_program?

Comment (Tansito) [reply client/quantum_serverless/core/decorators.py:349]: well, good question: it is `@distributed_program` or `@distribute_program` here? 😂

Comment (IceKhan13) [reply client/quantum_serverless/core/decorators.py:349]: we have `distribute_task`, therefore I'm using `distribute_program` :)

Comment (Tansito) [reply client/quantum_serverless/core/decorators.py:349]: well, that close the question quite easily 😂

Comment (akihikokuroda) [reply client/quantum_serverless/core/decorators.py:349]: Then, the `Example` must be match to the name :-)

File: client/quantum_serverless/core/decorators.py [hunk 2]
```diff
@@ -317,3 +321,116 @@ def wrapper(*args, **kwargs):
...
+    result = function(**arguments)
+    if result is not None:
+        save_result(result)
+"""
+
+
+def distribute_program(
+    provider: Optional[Any] = None,
+    dependencies: Optional[List[str]] = None,
+    working_dir: Optional[str] = None,
```
Comment (akihikokuroda) [inline client/quantum_serverless/core/decorators.py:344]: Multiple distributed programs that are executed at the same time with the same working directory may override the entry point file.

Comment (akihikokuroda) [reply client/quantum_serverless/core/decorators.py:344]: Adding random id to the entrypoint may be enough.

Comment (IceKhan13) [reply client/quantum_serverless/core/decorators.py:345]: Fixed in [c27d98f](https://github.com/Qiskit-Extensions/quantum-serverless/pull/908/commits/c27d98f8f7d6706d43f9629638104fd591188502)

File: docs/running/notebooks/06_running_programs_using_decorators.ipynb [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "f6fb114d-af27-4a89-83b4-b3806dfa8c1a",
+   "metadata": {},
+   "source": [
+    "# Running program using decorators [Experimental] \n",
+    "\n",
+    "In this tutorial we will describe alternative way (interface) of running your programs.\n",
+    "\n",
+    "> &#x26A0; This interface is experimental.\n",
```
Comment (psschwei) [inline docs/running/notebooks/06_running_programs_using_decorators.ipynb:12]: nit: I'd move the limitations here

File: docs/running/notebooks/06_running_programs_using_decorators.ipynb [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+   "id": "9f842194-b2f0-4887-a198-a7044689a9c0",
+   "metadata": {},
+   "source": [
+    "### Limitations\n",
+    "\n",
+    "This is experimental interfact, therefore it is subjected to breaking changes.\n",
+    "\n",
+    "For limitations of this interface:\n",
+    "1. functions decorated with distribute_program, can only accept named arguments for now. E.g do not use `my_program(argument1)`, instead specify name of the argument `my_program(argument1=argument1)`\n",
+    "2. function return will run `quantum_serverless.save_result` function under the hood, which means return values must be json serializable values in form of dictionary (with values as all Python native types, like strings, lists, dicts, `numpy` arrays, `QuantumCircuit`, `Operator`, etc.)\n",
```
[the remark this answers is not in the document]
Comment (IceKhan13) [reply docs/running/notebooks/06_running_programs_using_decorators.ipynb:286]: no, no returns means results will be `None` or empty

File: docs/running/notebooks/06_running_programs_using_decorators.ipynb (hunk 3 above)
Comment (IceKhan13) [reply docs/running/notebooks/06_running_programs_using_decorators.ipynb]: moved [c456a44](https://github.com/Qiskit-Extensions/quantum-serverless/pull/908/commits/c456a4485d93a00ee8fea50c5aba2d094a2624e4)

**Review by akihikokuroda — APPROVED** [review]
LGTM!  Thanks!

**Pull request merged**